### PR TITLE
[CIR] floating-point, pointer, and function types

### PR DESCRIPTION
**Pull request by dkolsen-pgi**
Upstream ClangIR support for `void` type, floating-point types, pointer types, and function types.

Floating-point support is missing the IBM double-double format, because that hasn't been implemented in the incubator project yet.

Pointer types do not yet support address spaces.

Function type support includes only the return type and the parameter types.  The many other properties and attributes of function types will be upstreamed later.

### CODE REVIEW

File: clang/include/clang/CIR/Dialect/IR/CIRTypes.td
Comment (efriedma-quic) [inline clang/include/clang/CIR/Dialect/IR/CIRTypes.td:206]: FYI, clang supports targets where "double" is 32 bits.  Granted, we're talking about pretty obscure targets, but maybe worth implementing now while you're looking at the relevant code.

File: clang/include/clang/CIR/Interfaces/CIRFPTypeInterface.td
Comment (erichkeane) [inline clang/include/clang/CIR/Interfaces/CIRFPTypeInterface.td:1]: Wasn't there some discussion on the last patch about the header not containing certain info?  I see this has the top level header, and the above file (CIRFPTypeInterface.h) having the description.  Which were we supposed to skip?

File: clang/lib/CIR/CodeGen/CIRGenTypes.cpp
Comment (erichkeane) [inline clang/lib/CIR/CodeGen/CIRGenTypes.cpp:33]: First, `tag` type, not `tagged` type, as this just represents any of the struct, class, union, or enum types.

Second, is the rest of this true?  Non-Tag types include atomics, Attributed types, bitint types, block pointers, complex, various dependent types, vector types, etc.  Do we really just mean to do `BuiltinType` here instead?  Or do we really mean we can do the rest already?

File: clang/lib/CIR/Dialect/IR/CIRTypes.cpp
Comment (erichkeane) [inline clang/lib/CIR/Dialect/IR/CIRTypes.cpp:375]: Why the close paren only in var-args?  What about in all other cases?

File: clang/include/clang/CIR/Dialect/IR/CIRTypes.td
Comment (dkolsen-pgi) [reply clang/include/clang/CIR/Dialect/IR/CIRTypes.td:206]: We'll have to tackle the more unusual and lesser-used targets, such as ones with 32-bit `double`, when we get there.  I can't get that code correct without having such a target to test on.

This effort is primarily upstreaming existing code from the ClangIR incubator project (which currently has limited target support) to the main LLVM repository.  I prefer to limit the amount of new code and new functionality as part of the upstreaming process.

File: clang/include/clang/CIR/Interfaces/CIRFPTypeInterface.td
Comment (dkolsen-pgi) [reply clang/include/clang/CIR/Interfaces/CIRFPTypeInterface.td:1]: You're right.  The updated guidelines say that this line doesn't need to contain any useful information.  I'll remove it.

File: clang/lib/CIR/CodeGen/CIRGenTypes.cpp
Comment (dkolsen-pgi) [reply clang/lib/CIR/CodeGen/CIRGenTypes.cpp:33]: This function is used to check if a type is problematic when it is used as part of a function type.  According to the comment here, only incomplete `class`/`struct`/`union` types are problematic when creating ClangIR function types.

The currently upstreamed ClangIR code can't handle atomic types or vector types or some of the other types you mentioned, in any context.  But when ClangIR can handle them, they will just work as part of function types; the code here won't need to change.

This entire function could have been left out of this PR since it doesn't do anything useful yet.  But because everything in the function except for the last line builds and runs, I decided to include it in this PR.

File: clang/lib/CIR/Dialect/IR/CIRTypes.cpp
Comment (dkolsen-pgi) [reply clang/lib/CIR/Dialect/IR/CIRTypes.cpp:375]: Printing the ')' is not in the varargs section.  It is at the top level of the function.  It is always output.

File: clang/include/clang/CIR/Dialect/IR/CIRTypes.td
Comment (erichkeane) [reply clang/include/clang/CIR/Dialect/IR/CIRTypes.td:206]: I wonder if we should have a 'not yet implemented' diagnostic that checks the size of 'double' to make sure it is 8?  That way we don't end up generating invalid code.  (Basically, check `ASTContext::getTypeSizeInChars` during setup?).

File: clang/include/clang/CIR/Interfaces/CIRFPTypeInterface.td
Comment (erichkeane) [reply clang/include/clang/CIR/Interfaces/CIRFPTypeInterface.td:1]: If you would, please audit the rest of the files that get added, I thought I noticed it a few places scrolling through.

File: clang/lib/CIR/CodeGen/CIRGenTypes.cpp
Comment (erichkeane) [reply clang/lib/CIR/CodeGen/CIRGenTypes.cpp:33]: In that case, I think just an assert in the  '!TagType' check here would be valuable, just that asserts it is a `BuiltinType` (or, the NYI diagnostic).

File: clang/lib/CIR/Dialect/IR/CIRTypes.cpp
Comment (erichkeane) [reply clang/lib/CIR/Dialect/IR/CIRTypes.cpp:375]: Coulda sworn I triple checked that before commenting.. Sorry about that.

File: clang/lib/CIR/CodeGen/CIRGenTypes.cpp
Comment (dkolsen-pgi) [reply clang/lib/CIR/CodeGen/CIRGenTypes.cpp:33]: I disagree about this one.  If the source code uses a vector type as the type of a function parameter, ClangIR code gen will fail with a NYI error in `CIRGenTypes::convertType` when trying to convert the Clang vector type into a CIR type.  There isn't a good reason to add a general-purpose check here in a function that is looking for a very specific condition that has only to do with function types.

Maybe this function should be renamed to `isConvertedTypeUsableInFunctionTypes`, to emphasize that it is not checking if a type is convertible, but checking whether or not a type is problematic if used as a return type or a parameter type within a function type.

File: clang/include/clang/CIR/Interfaces/CIRFPTypeInterface.td
Comment (dkolsen-pgi) [reply clang/include/clang/CIR/Interfaces/CIRFPTypeInterface.td:1]: Done.  `CIRFPTypeInterface.cpp` had the same issue, which I fixed in a soon-to-be-commited change.

I don't plan to make this change in any existing files that were added in earlier commits, only in new files going forward.

File: clang/include/clang/CIR/Dialect/IR/CIRTypes.td
Comment (dkolsen-pgi) [reply clang/include/clang/CIR/Dialect/IR/CIRTypes.td:206]: Good idea.  But I'm not sure where in the setup code to put such a check, since I need to check the format of the Clang types `float` and `double` and I'm not sure how best to access them in the ClangIR setup code.  Instead I have added `assert`s to the `BuiltinType::Float` and `BuiltinType::Double` cases in `CIRGenTypes::convertType`.  The `assert` will fire if the source code tries to use `float` or `double` on a platform where they aren't IEEE 32-bit and 64-bit.  The type `long double` already has a similar check in `CIRGenBuilderTy::getLongDoubleTy`.  I believe all the other non-standard floating-point types have fixed formats that shouldn't depend on the target, so they don't need similar checks.

Comment (erichkeane) [reply clang/include/clang/CIR/Dialect/IR/CIRTypes.td:206]: I think those asserts are good enough, thanks!

File: clang/include/clang/CIR/Interfaces/CIRFPTypeInterface.td
Comment (erichkeane) [reply clang/include/clang/CIR/Interfaces/CIRFPTypeInterface.td]: Yep, makes sense to me.

File: clang/lib/CIR/CodeGen/CIRGenTypes.cpp
Comment (erichkeane) [reply clang/lib/CIR/CodeGen/CIRGenTypes.cpp]: IMO, the assert here makes a lot of sense, in that it shows off the assumption that we're intentionally not handling these types.  The point is that we're NOT doing the work to figure out if those types are problematic if used in return /parameter in a function type (for good reason, since it'll fail earlier).  SO we should assert for that, so that when `CIRGenTypes::convertType` stops emitting the NYI, that this will be quickly noticed, rather than returning a value we wouldn't be sure of.

Comment (erichkeane) [reply clang/lib/CIR/CodeGen/CIRGenTypes.cpp]: >It is known that incomplete tag types are the only ones that are problematic in this context.

This seems inaccurate to me.  CIR doesn't seem to be able to represent(see here: https://github.com/llvm/clangir/blob/main/clang/include/clang/CIR/Dialect/IR/CIRTypes.td)  Atomic Types, BitInt types, Block types, and Matrix types upon quick look (that is perhaps an incomplete list).  

The function was written across; 
https://github.com/llvm/clangir/commit/101e732565d0e5004c5191b0053eebd8e849cade
and 
https://github.com/llvm/clangir/commit/93f55f6d43be394250009013c83d343be7c63f41

Neither of which mention those types specifically?  It seems it USED to have an assert for not-tag-type, though it isn't clear that the authors considered the above types.

Can @lanza and @bcardosolopes chime in and let me know if I'm misinterpreting the commits there?

**Comment by keryell** [comment]
@dkolsen-pgi  Great PR!
Since it touches functions returning `!void` or `!cir.void`, I made some changes recently with https://github.com/llvm/clangir/commit/568b51537e573c7a8fb616cda5cdd0aa54bc0832 to avoid disrupting some MLIR invariants.
Are you up-streaming the changes in commit order or are you up-streaming the changes with the latest version of a feature?
@joker-eph How to up-stream some changes which are cross-project like https://github.com/llvm/clangir/pull/1203 which introduces a new MLIR parser/pretty-printer feature like:
https://github.com/llvm/clangir/pull/1203/files#diff-cfea91beb87a3b28295e5612974f0556b4daf9a42c1adeb65e56f0dec062feaa
https://github.com/llvm/clangir/pull/1203/files#diff-6c1c134d749d53cf770548806b955401e47b1f15ffe084905a0b132d73b0b1fe
https://github.com/llvm/clangir/pull/1203/files#diff-dabff257abfd11fded480c2dfc0c2afdf640769384af9e37820e11003a75edb1
to allow some keywords starting with `!`?

File: clang/include/clang/CIR/Dialect/IR/CIRTypes.td
Comment (bcardosolopes) [reply clang/include/clang/CIR/Dialect/IR/CIRTypes.td:206]: Those asserts are probably the best way to go for now.

File: clang/lib/CIR/CodeGen/CIRGenTypes.cpp
Comment (bcardosolopes) [inline clang/lib/CIR/CodeGen/CIRGenTypes.cpp:180]: If we know there are possible 32 bit doubles (given previous comment), this looks more like NYI?

File: clang/lib/CIR/Dialect/IR/CIRTypes.cpp
Comment (bcardosolopes) [inline clang/lib/CIR/Dialect/IR/CIRTypes.cpp:391]: Can we get MissingFeature asserts for the ones here or below? Might be worth introducing that upstream if we haven't already?

**Review by bcardosolopes — APPROVED** [review]
LGTM with some minor nits

File: clang/lib/CIR/CodeGen/CIRGenTypes.cpp
Comment (bcardosolopes) [reply clang/lib/CIR/CodeGen/CIRGenTypes.cpp]: I think it's totally fine to be conservative about assumptions, we need asserts for both things that are NYI but also for things we assume as invariants, IMO the more the merrier cause it's easier to catch miscompilations.

> Neither of which mention those types specifically? It seems it USED to have an assert for not-tag-type, though it isn't clear that the authors considered the above types.

I don't remember the full context anymore, but assertion on a assumed invariant in the incubator to be even more protective also sounds good to me (guess running the tests upon such change will give us the final answer).

**Comment by bcardosolopes** [comment]
@keryell 

> How to up-stream some changes which are cross-project like https://github.com/llvm/clangir/pull/1203 which introduces a new MLIR parser/pretty-printer feature like:

My bad here, we usually don't accept such changes and I missed that during review time. ClangIR policy to changes outside CIR is to upstream them to MLIR and after that's done we cherry-pick it to the incubator (so rebases are smooth). I'll have to revert your change, sorry for the churn!

**Comment by keryell** [comment]
> My bad here, we usually don't accept such changes and I missed that during review time. ClangIR policy to changes outside CIR is to upstream them to MLIR and after that's done we cherry-pick it to the incubator (so rebases are smooth). I'll have to revert your change, sorry for the churn!

@bcardosolopes I can work on the phase 2 of the change I had in mind and remove also the `!void` from function type textual IR to avoid adding a new feature in the MLIR parsing infrastructure.

**Comment by bcardosolopes** [comment]
> @bcardosolopes I can work on the phase 2 of the change I had in mind and remove also the `!void` from function type textual IR to avoid adding a new feature in the MLIR parsing infrastructure.

Neat, if you can do that I won't need to revert and make it easier for everyone

File: clang/lib/CIR/Dialect/IR/CIRTypes.cpp
Comment (dkolsen-pgi) [reply clang/lib/CIR/Dialect/IR/CIRTypes.cpp:391]: You are right that a MissingFeature assert would be appropriate here.  The MissingFeature asserts are not in upstream yet.  I haven't felt an urge to migrate them yet because I find them confusing, and have to look at the code a couple times to understand what is happening.  In a (hopefully not-too-distant) future commit I will look at the MissingFeature asserts in the incubator and put something similar in upstream.  But not in this commit.

File: clang/lib/CIR/CodeGen/CIRGenTypes.cpp
Comment (dkolsen-pgi) [reply clang/lib/CIR/CodeGen/CIRGenTypes.cpp:180]: I am adding "NYI" to the assert messages.  But I like keeping these as asserts rather than compiler errors because `float` and `double` with unusual formats is not something we will support anytime soon, so the more drastic assertion failure seems more appropriate.

File: clang/lib/CIR/CodeGen/CIRGenTypes.cpp
Comment (dkolsen-pgi) [reply clang/lib/CIR/CodeGen/CIRGenTypes.cpp]: @bcardosolopes What exactly do you suggest doing here in this function.  Asserting on assumed invariants is great.  But I don't see any invariants here that could be asserted.

**Comment by dkolsen-pgi** [comment]
@keryell :
> Are you up-streaming the changes in commit order or are you up-streaming the changes with the latest version of a feature?

Upstreaming is happening in logical chunks, using the latest version of the feature.  I am paying no attention to the order of commits in the incubator project.  Unfortunately, your change landed after I started working on this change (or at least after I had last refreshed my incubator workspace).  So I didn't see it, and didn't incorporate your change into this one.

Once this PR is merged, could you please make your change again in the LLVM repository?  Bruno floated this idea in #clangir on Discord, of asking people who change already upstreamed code to make the change in both the incubator and the LLVM repository.  You may be the first guinea pig for this.  We can see if this is a good process.

File: clang/lib/CIR/CodeGen/CIRGenTypes.cpp
Comment (bcardosolopes) [reply clang/lib/CIR/CodeGen/CIRGenTypes.cpp]: I don't remember the details on why it was necessary at the time to deviate - so I just made this be equivalent to traditional codegen and everything passes: https://github.com/llvm/clangir/commit/04d7dcfb2582753f3eccbf01ec900d60297cbf4b

That said, I don't think we need to introduce any asserts here, can't get better than OG for the sake of completeness. Not supported types will crash given the other mechanisms David mentioned, but the recursive logic is sound with pre-existing work.

Unless Erich has any other concerns, this LGTM to land.

Comment (erichkeane) [reply clang/lib/CIR/CodeGen/CIRGenTypes.cpp]: If this includes llvm/clangir@04d7dcf in it, rather than the current implementation, I think I'm OK without any asserts.  I'll Approve, but conditional on that patch being included in this one.

**Review by erichkeane — APPROVED** [review]
Approve, conditional on isFuncParamTypeConvertible change being pulled from the incubator.

**Pull request merged**
